easybuild/tools/singularity/utilities.py

```diff
@@ -0,0 +1,352 @@
+			sys.exit(1)
+	else:
+		
+		print "Can't find image path ", bootstrap_imagepath
+		EasyBuildError("Can't find image path %s", bootstrap_imagepath)
+		sys.exit(1)
+
+    # if option is shub or docker		
+    else:
+	bootstrap_image = bootstrap_list[1]
+        image_tag = "NONE"
+    	# format --singularity-bootstrap shub:<image>:<tag>
+        if len(bootstrap_list) == 3:
+		image_tag = bootstrap_list[2]
+
+    module_scheme = get_module_naming_scheme()
+    
+    # bootstrap from local image
+    if bootstrap_type == "localimage":
+	bootstrap_content = "Bootstrap: " + bootstrap_type + " \n"
+	bootstrap_content += "From: " + bootstrap_imagepath + "\n" 
+    # default bootstrap is shub or docker
+    else:
+	    bootstrap_content = "BootStrap: " + bootstrap_type + "\n" 
+
+	    if image_tag == "NONE":
+		    bootstrap_content += "From: " + bootstrap_image  + "\n"
+	    else:
+		    bootstrap_content += "From: " + bootstrap_image + ":" + image_tag  + "\n"
+    
+    if module_scheme == "HierarchicalMNS":
+	    modulepath = "/app/modules/all/Core"
+    else:
+	    modulepath = "/app/modules/all/"
+
+    post_content = """
+%post
+"""
+    # if there is osdependencies in easyconfig then add them to Singularity recipe	
+    if len(osdeps) > 0:
+    	# format: osdependencies = ['libibverbs-dev', 'libibverbs-devel', 'rdma-core-devel']
+        if isinstance(osdeps[0],basestring):
+	     	for os_package in osdeps:
+		     	post_content += "yum install -y " + os_package + " || true \n"
+	# format: osdependencies = [('libibverbs-dev', 'libibverbs-devel', 'rdma-core-devel')]		
+	else:		
+	     	for os_package in osdeps[0]:
+		     	post_content += "yum install -y " + os_package + " || true \n"
+
+    post_content += "su - easybuild \n"
+ 
+    environment_content = """
+%environment
+source /etc/profile
+"""
+    
+    # check if toolchain is specified, that affects how to invoke eb and module load is affected based on module naming scheme
+    if tcname != "dummy":
+	# name of easyconfig to build
+        easyconfig  = appname + "-" + appver + "-" + tcname + "-" + tcver +  appversuffix + ".eb"
+	# name of Singularity defintiion file 
+        def_file  = "Singularity." + appname + "-" + appver + "-" + tcname + "-" + tcver +  appversuffix
+
+	ebfile = os.path.splitext(easyconfig)[0] + ".eb"
+        post_content += "eb " + ebfile  + " --robot --installpath=/app/ --prefix=/scratch --tmpdir=/scratch/tmp  --module-naming-scheme=" + module_scheme + "\n"
+
+	# This would be an example like running eb R-3.3.1-intel2017a.eb --module-naming-scheme=HierarchicalMNS. In HMNS you need to load intel/2017a first then R/3.3.1
+        if module_scheme == "HierarchicalMNS":
+                environment_content += "module use " + modulepath + "\n" 
+        	environment_content +=  "module load " + os.path.join(tcname,tcver) + "\n"
+        	environment_content +=  "module load " + os.path.join(appname,appver+appversuffix) + "\n"
```

Comment:
Does it make sense to do not load the module by default?
I am envisioning a usage where we can potentially pack different toolchains in order to empower user to compile their code. Not only to run applications.

Reply:
You need module load in %environment section so that when you do ``singularity shell`` or ``singularity exec`` it will load the module. See http://singularity.lbl.gov/docs-environment-metadata for more detail on %environment. 

I am still debating whether to support HMNS because this cause issue if you use one container with HMNS as bootstrap agent and then build another app with EasyBuildMNS this would cause two different module trees. @boegel what do you think

Reply:
It is possible to have a single installation path and multiple module naming schemes available since https://github.com/easybuilders/easybuild-framework/issues/1535 was resolved. It would require a bit of configuring though.